index.js

```diff
@@ -107,34 +103,75 @@ async function init() {
 
 	updateStatus('Retrieving directory info…');
 
-	await validateInput(repo);
+	const repoMetadata = await validateInput(`${user}/${repo}`);
+
+	const files = await listContent.viaTreesApi({
+		user,
+		repository: repo,
+		ref,
+		directory: decodeURIComponent(dir),
+		token: localStorage.token,
+		getFullData: true
+	});
 
-	const files = await listContent.viaTreesApi(`${repo}#${ref}`, decodeURIComponent(dir), localStorage.token);
 	if (files.length === 0) {
 		updateStatus('No files to download');
 		return;
 	}
 
-	updateStatus(`Downloading (0/${files.length}) files…`, '\n• ' + files.join('\n• '));
+	updateStatus(`Downloading (0/${files.length}) files…`, '\n• ' + files.map(file => file.path).join('\n• '));
 
 	let downloaded = 0;
 	const zip = new JSZip();
 	const controller = new AbortController();
-	const download = async path => {
-		const response = await fetch(`https://raw.githubusercontent.com/${repo}/${ref}/${path}`, {
+
+	const getBlobFromResponse = repoMetadata.private ?
+		async response => {
+			const {content, encoding} = await response.json();
+
+			if (encoding !== 'base64') {
+				throw new Error(`Unknown encoding "${encoding}"`);
+			}
+
+			let decodingResponse;
+			try {
+				decodingResponse = await fetch(`data:application/octet-stream;base64,${content}`);
+			} catch {
+				throw new Error('Decoding failed');
+			}
+
+			if (!decodingResponse.ok) {
+				throw new Error('Decoding failed');
+			}
+
+			return decodingResponse.blob();
+		} :
+		response => response.blob();
+
+	const fetchFile = repoMetadata.private ?
+		file => fetch(file.url, {
+			headers: {
+				Authorization: `Bearer ${localStorage.token}`
+			},
+			signal: controller.signal
+		}) :
+		file => fetch(`https://raw.githubusercontent.com/${user}/${repo}/${ref}/${file.path}`, {
 			signal: controller.signal
 		});
 
+	const download = async file => {
+		const response = await fetchFile(file);
+
 		if (!response.ok) {
-			throw new Error(`HTTP ${response.statusText} for ${path}`);
+			throw new Error(`HTTP ${response.statusText} for ${file.path}`);
 		}
 
-		const blob = await response.blob();
+		const blob = await getBlobFromResponse(response);
```

Comment:
You currently have 4 functions (`fetchFile` x 2, `getBlobFromResponse` x 2) that handle the download.

2 are enough, path goes in, blob comes out:

- `fetchPublicFile`
- `fetchPrivateFile`

Reply:
It was a tradeoff I considered. Had decided to not duplicate the error handling in between, but now that I see it it's actually barely 3 lines.